connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala

```diff
@@ -380,7 +396,26 @@ class SparkConnectPlanner(session: SparkSession) {
     }
     val attributes = structType.toAttributes
     val proj = UnsafeProjection.create(attributes, attributes)
-    new logical.LocalRelation(attributes, rows.map(r => proj(r).copy()).toSeq)
+    val relation = logical.LocalRelation(attributes, rows.map(r => proj(r).copy()).toSeq)
+
+    if (rel.hasDatatype || rel.hasDatatypeStr) {
+      // rename columns and update datatypes
+      val schema = if (rel.hasDatatype) {
+        DataTypeProtoConverter
+          .toCatalystType(rel.getDatatype)
+          .asInstanceOf[StructType]
+      } else {
+        parseDatatypeString(rel.getDatatypeStr)
+          .asInstanceOf[StructType]
+      }
+      Dataset
+        .ofRows(session, logicalPlan = relation)
+        .toDF(schema.names: _*)
```

Comment:
This throws away the names of nested fields. Perhaps we can do something like the following instead:
```scala
    val (rows, inferredSchema) = ArrowConverters.fromBatchWithSchemaIterator(
      Iterator(rel.getData.toByteArray),
      TaskContext.get())
    if (inferredSchema == null) {
      throw InvalidPlanInput(s"Input data for LocalRelation does not produce a schema.")
    }

    val schemaType = if (rel.hasDataType) {
      DataTypeProtoConverter.toCatalystType(rel.getDataType)
    } else if (rel.hasDataTypeString) {
      parseDatatypeString(rel.getDataTypeString)
    } else {
      inferredSchema
    }

    val schemaStruct = schemaType match {
      case s: StructType => s
      case d => StructType(Seq(StructField("value", d)))
    }

    val attributes = schemaStruct.toAttributes
    val proj = UnsafeProjection.create(attributes, attributes)
    new logical.LocalRelation(attributes, rows.map(r => proj(r).copy()).toSeq)
```

Reply:
we need to use `inferredSchema` in `proj`, otherwise it may fail due to datatype mismatch

Reply:
Will it work, if we use `val proj = UnsafeProjection.create(attributes, inferredSchema .toAttributes)` instead? We will likely also have to add some validation, similar to validation that pyspark currently performs using `_make_type_verifier`.

Reply:
still fails:

```
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.UNKNOWN
        details = "Couldn't find 0#1215L in [0#1219L,1#1220L,2#1221L,3#1222L]"
        debug_error_string = "{"created":"@1670584625.321243000","description":"Error received from peer ipv6:[::1]:15002","file":"src/core/lib/surface/call.cc","file_line":1064,"grpc_message":"Couldn't find 0#1215L in [0#1219L,1#1220L,2#1221L,3#1222L]","grpc_status":2}"
```

Reply:
seems due to unresolved